File: client/src/units.js

import store from './store/store'


export default {

    toML(otherAmount, otherUnits) {
        let units = store.state.units
        if (! (otherUnits in units.conversions))
            throw 'Invalid units: ' + otherUnits
        return otherAmount * units.conversions[otherUnits].toML
    },

    toOther(mlAmount, otherUnits) {
        let units = store.state.units
        if (! (otherUnits in units.conversions))
            throw 'Invalid units: ' + otherUnits
        return mlAmount / units.conversions[otherUnits].toML
    },
    
    format(otherAmount, otherUnits, appendUnits = true) {
        let units = store.state.units
        if (! (otherUnits in units.conversions))
            return ''
            //throw 'Invalid units: ' + otherUnits
        let str = otherAmount.toFixed(units.conversions[otherUnits].precision)
        if (appendUnits)
            str += ' ' + otherUnits
        return str
    },
    
    defaultUnits() {
        let units = store.state.units
        return units['default']
    },
    
    units() {
        let units = store.state.units
        return units['order']
    },
    
}

File: server/barbot/units.py

from .bus import bus
from .config import config

default = ''
order = []
conversions = {}


@bus.on('server/start1')
def _bus_serverStart():
    load()

@bus.on('config/loaded')
def _bus_configLoaded():
    load()

def load():
    global default, order, conversions
    for (k, v) in config['units'].items():
        if k == 'default':
            default = v
        elif k == 'order':
            order = v.split(',')
        else:
            v = v.split(',')
            conversions[k] = {
                'label': k,
                'toML': float(v[0]),
                'precision': int(v[1])
            }

def toML(otherAmount, otherUnits):
    if otherUnits not in conversions:
        raise ValueError('Invalid units: {}'.format(otherUnits))
    return otherAmount * conversions[otherUnits]['toML']

def toOther(mlAmount, otherUnits):
    if otherUnits not in conversions:
        raise ValueError('Invalid units: {}'.format(otherUnits))
    return mlAmount / conversions[otherUnits]['toML']
    
def format(otherAmount, otherUnits, appendUnits = True):
    if otherUnits not in conversions:
        raise ValueError('Invalid units: {}'.format(otherUnits))
    str = format(otherAmount, '.{}f'.format(conversions[otherUnits]['precision']))
    if appendUnits:
        str = str + ' ' + otherUnits
    return str

def export():
    return {
        'default': default,
        'order': order,
        'conversions': conversions,
    }
